tensorflow/contrib/periodic_intersperse/core/kernels/periodic_intersperse_op.h

```diff
@@ -0,0 +1,231 @@
+// =============================================================================
+// Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#ifndef TENSORFLOW_KERNELS_PERIODICINTERSPERSE_OP_H_
+#define TENSORFLOW_KERNELS_PERIODICINTERSPERSE_OP_H_
+
+#include <cmath>
+#include <type_traits>
+#include <vector>
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/shape_inference.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+
+template <class IndexVecT, class IndexT>
+IndexT compute_input_index(
+    IndexVecT& Y, const IndexT& i, const IndexVecT& X, const int& q,
+    const std::vector<tensorflow::int64>& g,
+    const std::vector<tensorflow::int64>& G, IndexT& result,
+    std::vector<IndexT>& output_indices,
+    const typename std::decay<decltype(Y.size())>::type& rank) {
+  result = 0;
+  output_indices.clear();
+
+  // un-rasterize the output index
+  {
+    auto last_reduced_i = i;
+    auto r = rank;
+    do {
```

Comment:
I changed the type of rank to be signed so that the loop could be written as you suggested.